File: Src/java/cql-to-elm/src/test/java/org/cqframework/cql/cql2elm/matchers/Quick2DataType.java
package org.cqframework.cql.cql2elm.matchers;

import org.hamcrest.Description;
import org.hamcrest.Factory;
import org.hamcrest.Matcher;
import org.hamcrest.TypeSafeDiagnosingMatcher;

import javax.xml.namespace.QName;

public class Quick2DataType extends TypeSafeDiagnosingMatcher<QName> {
    private QName expectedValue;

    public Quick2DataType(String fullName) {
        super();

        expectedValue = new QName("http://hl7.org/fhir/us/qicore", fullName, "");
    }

    @Override
    protected boolean matchesSafely(QName item, Description mismatchDescription) {
        if (! expectedValue.getNamespaceURI().equals(item.getNamespaceURI())) {
            mismatchDescription.appendText("had wrong namespace: ").appendValue(item.getNamespaceURI());
            return false;
        }

        if (! expectedValue.getLocalPart().equals(item.getLocalPart())) {
            mismatchDescription.appendText("had wrong local part: ").appendText(item.getLocalPart());
            return false;
        }

        if (! expectedValue.getPrefix().equals(item.getPrefix())) {
            mismatchDescription.appendText("had wrong prefix: ").appendText(item.getPrefix());
            return false;
        }

        return true;
    }

    @Override
    public void describeTo(Description description) {
        description.appendValue(expectedValue);
    }

    @Factory
    public static <T> Matcher<QName> quick2DataType(String fullName) {
        return new Quick2DataType(fullName);
    }

    @Factory
    public static <T> Matcher<QName> quick2DataType(String topic, String modality) {
        return new Quick2DataType(topic + modality + "Occurrence");
    }

    @Factory
    public static <T> Matcher<QName> quick2DataType(String topic, String modality, String occurrence) {
        return new Quick2DataType(topic + modality + occurrence);
    }
}

File: Src/java/cql-to-elm/src/jvmTest/kotlin/org/cqframework/cql/cql2elm/matchers/Quick2DataType.kt
package org.cqframework.cql.cql2elm.matchers

import javax.xml.namespace.QName
import org.hamcrest.Description
import org.hamcrest.Factory
import org.hamcrest.Matcher
import org.hamcrest.TypeSafeDiagnosingMatcher

class Quick2DataType(fullName: String) : TypeSafeDiagnosingMatcher<QName>() {
    private val expectedValue: QName = QName("http://hl7.org/fhir/us/qicore", fullName, "")

    @Suppress("ReturnCount")
    override fun matchesSafely(item: QName, mismatchDescription: Description): Boolean {
        if (expectedValue.namespaceURI != item.namespaceURI) {
            mismatchDescription.appendText("had wrong namespace: ").appendValue(item.namespaceURI)
            return false
        }

        if (expectedValue.localPart != item.localPart) {
            mismatchDescription.appendText("had wrong local part: ").appendText(item.localPart)
            return false
        }

        if (expectedValue.prefix != item.prefix) {
            mismatchDescription.appendText("had wrong prefix: ").appendText(item.prefix)
            return false
        }

        return true
    }

    override fun describeTo(description: Description) {
        description.appendValue(expectedValue)
    }

    @Suppress("MemberNameEqualsClassName")
    companion object {
        @JvmStatic
        @Factory
        fun quick2DataType(fullName: String): Matcher<QName?> {
            return Quick2DataType(fullName)
        }

        @JvmStatic
        @Factory
        fun quick2DataType(topic: String?, modality: String?): Matcher<QName?> {
            return Quick2DataType(topic + modality + "Occurrence")
        }

        @JvmStatic
        @Factory
        fun quick2DataType(
            topic: String?,
            modality: String?,
            occurrence: String?
        ): Matcher<QName?> {
            return Quick2DataType(topic + modality + occurrence)
        }
    }
}
